### PR TITLE
Espanso: Fix broken module to be compatible with Espanso version 2.x

### PR DESCRIPTION
**Pull request by liyangau**
### Description

Espanso module does not work with 2.x version.  In the new version configuration and matches are put into different folders as below.
```
config/
  default.yml
match/
  base.yml
```
The updated module supports creating multiple files for both config and matches. Users have the full flexibility to create multiple files for different apps.

### Checklist

- [ ] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

### CODE REVIEW

File: modules/services/espanso.nix [hunk 1]
```diff
@@ -25,41 +21,58 @@ in {
         defaultText = literalExpression "pkgs.espanso";
       };
 
-      settings = mkOption {
+      configs = mkOption {
         type = yaml.type;
-        default = { matches = [ ]; };
+        default = { default = { }; };
```
Comment (ncfavier) [inline modules/services/espanso.nix:26]: Do we need a default?

File: modules/services/espanso.nix [hunk 2]
```diff
@@ -25,41 +21,58 @@ in {
...
         '';
-        description = ''
-          The Espanso configuration to use. See
-          <link xlink:href="https://espanso.org/docs/configuration/"/>
-          for a description of available options.
+      };
+
+      matches = mkOption {
+        type = yaml.type;
+        default = { default.matches = [ ]; };
```
Comment (ncfavier) [inline modules/services/espanso.nix:47]: Ditto

File: modules/services/espanso.nix [hunk 3]
```diff
@@ -70,8 +83,16 @@ in {
...
-      yaml.generate "espanso-default.yml" cfg.settings;
+    xdg.configFile = let
+      configFiles = lib.mapAttrs' (name: value: {
+        name = "espanso/config/${name}.yml";
+        value = { source = yaml.generate "${name}.yml" value; };
+      }) config.services.espanso.configs;
+      matchesFiles = lib.mapAttrs' (name: value: {
+        name = "espanso/match/${name}.yml";
+        value = { source = yaml.generate "${name}.yml" value; };
+      }) config.services.espanso.matches;
```
Comment (ncfavier) [inline modules/services/espanso.nix:104]:
```suggestion
      }) cfg.configs;
      matchesFiles = lib.mapAttrs' (name: value: {
        name = "espanso/match/${name}.yml";
        value = { source = yaml.generate "${name}.yml" value; };
      }) cfg.matches;
```

File: modules/services/espanso.nix (hunk 1 above)
Comment (liyangau) [reply modules/services/espanso.nix:26]: I am following previous config here. I guess there is no harm to set an empty default when the users do not provide anything.

File: modules/services/espanso.nix (hunk 2 above)
Comment (liyangau) [reply modules/services/espanso.nix:47]: same as above

File: modules/services/espanso.nix [hunk 4]
```diff
@@ -25,53 +27,96 @@ in {
...
-                replace = "{{output}}";
-                vars = [{
-                  name = "output";
-                  type = "shell";
-                  params = { cmd = "echo Hello from your shell"; };
-                }];
-              }
-            ];
-          }
+          configs = {
```
Comment (bobvanderlinden) [inline modules/services/espanso.nix:34]: The example of configs shouldnt include the configs key.

**Review by bobvanderlinden — COMMENTED** [review]
Haven't tried to run this, but code looks good overal 👍 small suggestion.

File: modules/services/espanso.nix (hunk 1 above)
Comment (ncfavier) [reply modules/services/espanso.nix:32]: Oh, sorry, I was unclear. I meant: do we need a `default` that's not just `{}`? I.e., why the `default` inside the `default`?

Comment (liyangau) [reply modules/services/espanso.nix:33]: I use `default = { }; inside default is to create an empty `~/.config/espanso/config/default.yml` when it is not set.

Without it, Espanso will throw below error.

```
[ERROR] unable to load config

Caused by:
    missing config directory
```

**Review by ncfavier — APPROVED** [review]

**Comment by liyangau** [comment]
Looks like _pkgs.espanso_ does not `passthru.test` version. Shall I remove the checks of `versionAtLeast espansoVersion "2";` to move this forward?

**Comment by ncfavier** [comment]
The stubs don't have versions by default. I'm [working on it](https://github.com/nix-community/home-manager/pull/4069).

**Comment by ncfavier** [comment]
Should be fixed now if you rebase on master.

File: modules/services/espanso.nix [hunk 5]
```diff
@@ -1,19 +1,22 @@
...
-
 in {
-  meta.maintainers = with maintainers; [ lucasew ];
-
+  imports = [
+    (mkRemovedOptionModule [ "services" "espanso" "settings" ]
+      "Use services.espanso.configs and services.espanso.matches instead.")
+  ];
+  meta.maintainers =
+    [ lib.hm.maintainers.liyangau maintainers.bobvanderlinden ];
```
Comment (ncfavier) [inline modules/services/espanso.nix:19]: Why the change of maintainer?

File: modules/services/espanso.nix [hunk 6]
```diff
@@ -25,53 +28,96 @@ in {
...
   };
 
   config = mkIf cfg.enable {
-    assertions = [ (assertPlatform "services.espanso" pkgs platforms.linux) ];
+    assertions = [
+      (assertPlatform "services.espanso" pkgs platforms.linux)
+      {
+        assertion = versionAtLeast espansoVersion "2";
+        message = ''
+          Only support version Espanso version 2
```
Comment (ncfavier) [inline modules/services/espanso.nix:104]:
```suggestion
          The services.espanso module only supports Espanso version 2 or later.
```

File: modules/services/espanso.nix (hunk 5 above)
Comment (liyangau) [reply modules/services/espanso.nix:19]: I copied the maintainer part from another file.  I fix it with the correct ones...

Comment (liyangau) [reply modules/services/espanso.nix:19]: Speaking of maintainer, do we keep adding new maintainer in a list or in this case, override it so people know who writes the current one?

Comment (ncfavier) [reply modules/services/espanso.nix:19]: Presumably @lucasew still wants to be a maintainer for the module?

File: modules/services/espanso.nix [hunk 7]
```diff
@@ -1,19 +1,25 @@
...
 in {
-  meta.maintainers = with maintainers; [ lucasew ];
-
+  imports = [
+    (mkRemovedOptionModule [ "services" "espanso" "settings" ]
+      "Use services.espanso.configs and services.espanso.matches instead.")
+  ];
+  meta.maintainers = [
+    maintainers.lucasew
+    maintainers.bobvanderlinden
```
Comment (ncfavier) [inline modules/services/espanso.nix:20]: Does @bobvanderlinden want to be added?

Comment (liyangau) [reply modules/services/espanso.nix:20]: Yes,  he said it was fine. https://discourse.nixos.org/t/how-to-override-service-espanso-module-for-home-manager/28551/12

**Review by ncfavier — APPROVED** [review]
